### PR TITLE
feat: Separate business logic and database implementation details 

### PR DESCRIPTION
**Pull request by VikashKothary**

### CODE REVIEW

File: modules/core/src/main/java/org/reciplease/repository/IngredientRepository.java [hunk 1]
```diff
@@ -1,11 +1,15 @@
...
+import java.util.Optional;
 import java.util.UUID;
 
-public interface IngredientRepository extends JpaRepository<Ingredient, UUID> {
+public interface IngredientRepository {
+    List<Ingredient> findAll();
+    Optional<Ingredient> findById(final UUID uuid);
+    Ingredient save(final Ingredient ingredient);
+    List<Ingredient> saveAll(final List<Ingredient> ingredients);
     List<Ingredient> findByNameContains(String searchName);
```
Comment (rhys-saldanha) [inline modules/core/src/main/java/org/reciplease/repository/IngredientRepository.java]: Can we rename to `searchByName` please

File: modules/core/src/main/java/org/reciplease/repository/InventoryRepository.java [hunk 2]
```diff
@@ -1,14 +1,16 @@
...
 import java.util.List;
+import java.util.Optional;
 import java.util.UUID;
 
-public interface InventoryRepository extends JpaRepository<InventoryItem, UUID> {
+public interface InventoryRepository {
+    Optional<InventoryItem> findById(final UUID uuid);
+    InventoryItem save(final InventoryItem item);
+    List<InventoryItem> findAll();
     List<InventoryItem> findByExpirationIsGreaterThanEqual(LocalDate now);
```
Comment (rhys-saldanha) [inline modules/core/src/main/java/org/reciplease/repository/InventoryRepository.java]: Another clunky name, maybe `expiresAfter`?

File: modules/core/src/main/java/org/reciplease/repository/PlannedRecipeRepository.java [hunk 3]
```diff
@@ -1,12 +1,11 @@
...
-import org.springframework.data.jpa.repository.JpaRepository;
 
 import java.time.LocalDate;
 import java.util.List;
-import java.util.UUID;
 
-public interface PlannedRecipeRepository extends JpaRepository<PlannedRecipe, UUID> {
+public interface PlannedRecipeRepository {
+    PlannedRecipe save(PlannedRecipe plannedRecipe);
     List<PlannedRecipe> findByDateIsBetween(LocalDate start, LocalDate end);
```
Comment (rhys-saldanha) [inline modules/core/src/main/java/org/reciplease/repository/PlannedRecipeRepository.java:10]: `betweenDates`

File: modules/core/src/main/java/org/reciplease/repository/RecipeRepository.java [hunk 4]
```diff
@@ -1,9 +1,13 @@
...
 
+import java.util.List;
+import java.util.Optional;
 import java.util.UUID;
 
-public interface RecipeRepository extends JpaRepository<Recipe, UUID> {
+public interface RecipeRepository {
+    List<Recipe> findAll();
+    Recipe save(final Recipe recipe);
+    Optional<Recipe> findByUuid(final UUID uuid);
```
Comment (rhys-saldanha) [inline modules/core/src/main/java/org/reciplease/repository/RecipeRepository.java:12]: All the other repositories use ID over UUID, need to pick one for consistency. I prefer this.

File: modules/database/src/main/java/org/reciplease/model/BaseEntity.java [hunk 5]
```diff
@@ -1,13 +1,13 @@
 package org.reciplease.model;
 
 import io.swagger.v3.oas.annotations.media.Schema;
```
Comment (rhys-saldanha) [inline modules/database/src/main/java/org/reciplease/model/BaseEntity.java]: Hibernate entities don't need OAS annotations, I'm not happy that this package is even accessible in the database module.

File: modules/parent/pom.xml [hunk 6]
```diff
@@ -24,6 +24,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <java.version>11</java.version>
         <springdoc.version>1.5.12</springdoc.version>
+        <kotlin.version>1.6.10</kotlin.version>
```
Comment (rhys-saldanha) [inline modules/parent/pom.xml:27]: nice

File: modules/web/src/main/java/org/reciplease/dto/IngredientDto.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+import javax.validation.constraints.NotNull;
+import java.util.UUID;
+
+@Builder
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@JsonInclude(Include.NON_EMPTY)
+public class IngredientDto {
+    @Nullable
```
Comment (rhys-saldanha) [inline modules/web/src/main/java/org/reciplease/dto/IngredientDto.java]: Is it?...

**Review by rhys-saldanha — CHANGES_REQUESTED** [review]

**Comment by rhys-saldanha** [comment]
Code coverage is a bit lacking 😞

File: modules/core/src/main/java/org/reciplease/repository/IngredientRepository.java (hunk 1 above)
Comment (VikashKothary) [reply modules/core/src/main/java/org/reciplease/repository/IngredientRepository.java]: Done.

File: modules/core/src/main/java/org/reciplease/repository/InventoryRepository.java (hunk 2 above)
Comment (VikashKothary) [reply modules/core/src/main/java/org/reciplease/repository/InventoryRepository.java]: Done.

File: modules/parent/pom.xml (hunk 6 above)
Comment (VikashKothary) [reply modules/parent/pom.xml:27]: :P

File: modules/web/src/main/java/org/reciplease/dto/IngredientDto.java (hunk 7 above)
Comment (VikashKothary) [reply modules/web/src/main/java/org/reciplease/dto/IngredientDto.java]: No idea, so I'll just remove the annotation.

File: modules/core/src/main/java/org/reciplease/repository/PlannedRecipeRepository.java (hunk 3 above)
Comment (VikashKothary) [reply modules/core/src/main/java/org/reciplease/repository/PlannedRecipeRepository.java:10]: Done.

**Comment by VikashKothary** [comment]
We can definitely improve the code coverage, I'm just concerned how chunky this PR is becoming.

File: modules/core/src/main/java/org/reciplease/repository/RecipeRepository.java (hunk 4 above)
Comment (VikashKothary) [reply modules/core/src/main/java/org/reciplease/repository/RecipeRepository.java:12]: Totally agree. Done.

**Pull request merged**